### PR TITLE
Remove Required attribute from GenerateNuSpec parameters.

### PR DESCRIPTION
**Pull request by williamb1024**
Issue #57 

Removes Required attribute from GenerateNuSpec parameters. Enforces required parameters only if a template nuspec file is not specified.

This allows the Id, Version, Title, Authors, Owners, and Description attributes to be specified in the nuspec file without requiring the same value be placed in the nuproj file properties.

### CODE REVIEW

**Comment by williamb1024** [comment]
Issue #57 

Modified nuproj.targets to initialize the Id and Version properties from the nuspec target file, if the values are not defined within the nuproj file. 

I've broken something in the tests though..

Issue #59 

Added a clean cache (generally a cut & paste of the functionality defined in Microsoft.Common.CurrentVersion.targets from MSBuild).

**Comment by kovalikp** [comment]
I've checked your working branch. It seems you forgot to specify template property in `Task_GenerateNuSpec_NoRequiredParameters\NuGetPackage\NuGetPackage.nuproj`. Can you try to add there:

```xml
<PropertyGroup>
  <NuSpecTemplate>template.nuspec</NuSpecTemplate>
</PropertyGroup>
```

**Comment by williamb1024** [comment]
The nuspec template isn't supposed to be in the project. One of the test ensures that the build script will fail if the required parameters are not specified in either the template or the project.

Which tests do you see failing? All pass for me locally.

Sent from my iPhone

On Jan 17, 2015, at 2:26 PM, Pavol Kovalik <notifications@github.com<mailto:notifications@github.com>> wrote:


I've checked your working branch. It seems you forgot to specify template property in Task_GenerateNuSpec_NoRequiredParameters\NuGetPackage\NuGetPackage.nuproj. Can you try to add there:

<PropertyGroup>
  <NuSpecTemplate>template.nuspec</NuSpecTemplate>
</PropertyGroup>

File: src/NuProj.Tests/GenerateNuSpecTests.cs [hunk 1]
```diff
@@ -197,5 +198,81 @@ public void Task_GenerateNuSpec_UseInputFileName()
...
+            var solutionDir = Assets.GetScenarioDirectory(scenarioName);
+            var projectPath = Path.Combine(solutionDir, projectToBuild);
+
+            Dictionary<string, string> properties = new Dictionary<string, string>()
+            {
+                {"Id", "NoRequiredParameters"}
+            };
+
+            var result = await MSBuild.ExecuteAsync(projectPath, "GenerateNuSpec", properties);
+            result.AssertError("The 'Version' property cannot be empty or undefined.");
```
Comment (kovalikp) [inline src/NuProj.Tests/GenerateNuSpecTests.cs]: I'm getting `The "GenerateNuSpec" task was not given a value for the required parameter "Id".` here.

File: src/NuProj.Tests/GenerateNuSpecTests.cs [hunk 2]
```diff
@@ -197,5 +198,81 @@ public void Task_GenerateNuSpec_UseInputFileName()
...
+                {"NuSpecTemplate", templatePath}
+            };
+
+            // load the template manifest so we can compare it later..
+            Manifest templateManifest;
+            using (var inputStream = File.OpenRead(templatePath))
+                templateManifest = Manifest.ReadFrom(inputStream, false);
+
+            var result = await MSBuild.ExecuteAsync(projectPath, "GenerateNuSpec", properties);
+            result.AssertSuccessfulBuild();
```
Comment (kovalikp) [inline src/NuProj.Tests/GenerateNuSpecTests.cs]: Also getting `The "GenerateNuSpec" task was not given a value for the required parameter "Id".` here.

File: src/NuProj.Tests/GenerateNuSpecTests.cs [hunk 3]
```diff
@@ -197,5 +198,81 @@ public void Task_GenerateNuSpec_UseInputFileName()
...
+        }
+
+        [Theory]
+        [InlineData("Task_GenerateNuSpec_NoRequiredParameters", @"NuGetPackage\NuGetPackage.nuproj")]
+        public async System.Threading.Tasks.Task Task_GenerateNuSpec_NoRequiredParameters_IdOnly(string scenarioName, string projectToBuild)
+        {
+            var solutionDir = Assets.GetScenarioDirectory(scenarioName);
+            var projectPath = Path.Combine(solutionDir, projectToBuild);
+
+            Dictionary<string, string> properties = new Dictionary<string, string>()
```
Comment (kovalikp) [inline src/NuProj.Tests/GenerateNuSpecTests.cs]: You need to add default properties here. Use 
```csharp
var properties = MSBuild.Properties.Default.Add("Id", "NoRequiredParameters");
```

**Comment by kovalikp** [comment]
Ok, found the problem: you're not setting `NuProjPath`, `NuProjTasksPath` and other properties. This means NuProj will try to load everything from default location, most likely `$(MSBuildExtensionsPath)\NuProj\`. This is why tests and others people tests were failing. See comment to code how to fix it.

**Comment by williamb1024** [comment]
OK, the tests are corrected. Can you accept both issues in a single pull request, or should I split them into two different requests?

**Comment by kovalikp** [comment]
I can't. That's up to @terrajobst .

**Pull request closed without merging**